### PR TITLE
fix: preserve timeout without pooled context reuse

### LINKED ISSUE #966: 高并发下，某些请求会得到多个响应结果
### Before feedback (在反馈之前)

- [x] I've searched the existing Issues, Discussions and Google (我已经搜索了现有的 Issues, Discussions 和 Google)
- [x] The problem can be stably reproduced (这个问题可以被稳定复现)
- [ ] The problem is generated after upgrading (问题是在升级之后产生的)

### Operating System (操作系统)

Linux

### Golang Version (Go 版本)

1.25.5

### Goravel Version (Goravel 版本)

1.17.2

### Describe The Problem (描述问题)

在项目并发较高的时候，特别是有客户端因为超时断开了连接以后，后面的请求可能会返回多个返回内容。
某些内容都不是这次请求返回的内容，可能是之前哪个请求的返回内容，多个内容会拼接到一起像我这个日志记录一样
[2026-05-20 16:26:56.704] local.info: [HTTP] | Proto:HTTP/1.1 | PID:35675 | PORT:8090 | 200 | 4.209993ms | 172.18.0.1 | GET http://api.dsp.local/ping Headers:{"Accept-Encoding":["gzip"],"Connection":["close"],"Content-Type":["text/html"],"User-Agent":["hey/0.0.1"],"X-Forwarded-For":["172.18.0.1"],"X-Forwarded-Proto":["http"],"X-Real-Ip":["172.18.0.1"],"X-Request-Id":["02663b00c669fe81cbbca3e9ecd47a0e"]} | Headers:{"Content-Type":["application/json; charset=utf-8"]} Body:{"code":0,"message":"OK","data":[],"errors":[]}{"code":0,"message":"OK","data":[],"errors":[]}不需要复杂的代码，只需要使用接口耗时加长一些就能复现。

### Reproduction Code (复现代码或截图)

```go
```

### PR DESCRIPTION
**Pull request by hwbrzzl**
## Summary
- Keep automatic request timeouts while replacing the unsafe custom timeout runner with `gin-contrib/timeout` internally.
- Preserve `408 Request Timeout` responses while preventing late writes from a timed-out request from leaking into the next request.
- Add regression coverage for timeout completion ordering and request isolation after a timeout.

Closes https://github.com/goravel/goravel/issues/966

## Why
The previous timeout middleware enforced timeouts by running the Gin chain in a background goroutine and returning as soon as the deadline fired. That made Goravel return pooled request and response wrappers before the timed-out handler had actually stopped, which could let late writes bleed into a later request.

This change keeps the timeout feature instead of removing it. Goravel still exposes the same `Timeout(...)` middleware and still returns HTTP 408 on timeout, but the implementation now delegates to `gin-contrib/timeout`, which buffers the response, suppresses late writes after timeout, and waits for the worker goroutine to finish before control returns to Gin's pooled context lifecycle.

```go
facades.Route().Get("/ping", func(ctx http.Context) http.Response {
	select {
	case <-time.After(5 * time.Second):
		return ctx.Response().Success().Json(http.Json{
			"code":    0,
			"message": "OK",
			"data":    []any{},
			"errors":  []any{},
		})
	case <-ctx.Done():
		return nil
	}
})
```

### CODE REVIEW

File: context.go
Comment (hwbrzzl) [inline context.go:1]: Are the changes still needed?

File: middleware_timeout.go [hunk 1]
```diff
@@ -4,42 +4,30 @@ import (
...
-		}()
-
-		select {
-		case <-done:
-		case <-timeoutCtx.Done():
-			if errors.Is(timeoutCtx.Err(), context.DeadlineExceeded) {
-				ctx.Request().Abort(contractshttp.StatusRequestTimeout)
-			}
+			return
 		}
```
Comment (hwbrzzl) [inline middleware_timeout.go]: Is it necessary? ctx must be `*Context`.

File: middleware_timeout_test.go [hunk 2]
```diff
@@ -27,18 +27,50 @@ func TestTimeoutMiddleware(t *testing.T) {
...
+
+		select {
+		case <-done:
+		case <-time.After(time.Second):
+			t.Fatal("request did not complete after the handler returned")
+		}
 
-		route.ServeHTTP(w, req)
 		assert.Equal(t, contractshttp.StatusRequestTimeout, w.Code)
+		assert.Equal(t, http.StatusText(contractshttp.StatusRequestTimeout), w.Body.String())
```
Comment (hwbrzzl) [inline middleware_timeout_test.go:81]:
```suggestion
		assert.Equal(t, "timeout", w.Body.String())
```

File: middleware_timeout_test.go [hunk 3]
```diff
@@ -55,6 +87,62 @@ func TestTimeoutMiddleware(t *testing.T) {
...
+		close(allowReturn)
+
+		select {
+		case <-firstDone:
+		case <-time.After(time.Second):
+			t.Fatal("timed out request did not complete after the handler returned")
+		}
+
+		assert.Equal(t, contractshttp.StatusRequestTimeout, firstWriter.Code)
+		assert.Equal(t, http.StatusText(contractshttp.StatusRequestTimeout), firstWriter.Body.String())
```
Comment (hwbrzzl) [inline middleware_timeout_test.go:152]:
```suggestion
		assert.Equal(t, "timeout", firstWriter.Body.String())
```

File: context.go
Comment (hwbrzzl) [reply context.go:1]: Yes. The adapter swap no longer depends on the old `context.go` behavior for timeout handling, but these methods still need to be nil-safe because `Background()` and other test-created contexts can have a nil `Request`. I pushed a fix that initializes a default request before reading `Context()/Deadline()/Done()/Err()/Value()`.

File: middleware_timeout_test.go (hunk 3 above)
Comment (hwbrzzl) [reply middleware_timeout_test.go:152]: Same here: the runtime timeout body is `Request Timeout`, not `timeout`. I kept the assertion aligned with the actual middleware behavior and confirmed it with the added integration test in `goravel/example`.

File: middleware_timeout_test.go (hunk 2 above)
Comment (hwbrzzl) [reply middleware_timeout_test.go:81]: I kept the assertion against `http.StatusText(http.StatusRequestTimeout)` intentionally. The actual middleware response body here is the standard `Request Timeout`, and I also verified that behavior against a real integration test in `goravel/example` after wiring this repo via `replace github.com/goravel/gin => ../gin`.

File: middleware_timeout.go (hunk 1 above)
Comment (hwbrzzl) [reply middleware_timeout.go]: Yes, and I simplified it accordingly in `e1805c7`. `Timeout(...)` now calls `timeoutMiddleware(ctx.(*Context).Instance())` directly, since this driver always passes `*Context` through the middleware adapter.

File: context.go [hunk 4]
```diff
@@ -99,15 +99,15 @@ func (c *Context) Context() context.Context {
 }
 
 func (c *Context) Deadline() (deadline time.Time, ok bool) {
-	return c.instance.Deadline()
+	return c.instance.Request.Context().Deadline()
 }
 
 func (c *Context) Done() <-chan struct{} {
```
[the remark this answers is not in the document]
Comment (hwbrzzl) [reply context.go:112]: Fixed in `e1805c7`. `Background()` now installs a default request, and the context accessors call `ensureRequest()` before touching `Request.Context()` so they stay nil-safe for test-created contexts.

File: middleware_timeout_test.go (hunk 2 above)
Comment (hwbrzzl) [reply middleware_timeout_test.go:81]: But `ctx.Response().Status(contractshttp.StatusRequestTimeout).String("timeout")` is actually returned, why assert `http.StatusText(http.StatusRequestTimeout)`?

File: context.go [hunk 5]
```diff
@@ -130,3 +138,9 @@ func (c *Context) getGoravelContextValues() map[any]any {
 	}
 	return make(map[any]any)
 }
+
+func (c *Context) ensureRequest() {
+	if c.instance.Request == nil {
+		c.instance.Request = httptest.NewRequest(nethttp.MethodGet, "/", nil)
+	}
+}
```
Comment (hwbrzzl) [inline context.go]: It is unnecessary, remove it.

Comment (hwbrzzl) [reply context.go]: Agreed. I removed `ensureRequest()` in `bf5fad2` and moved the nil-request initialization into `NewContext`, so we keep the nil-safety without an extra helper or repeated checks in every accessor.

File: middleware_timeout_test.go (hunk 2 above)
Comment (hwbrzzl) [reply middleware_timeout_test.go:81]: Because that `String("timeout")` call happens after `<-ctx.Done()` inside the worker goroutine, and `gin-contrib/timeout` discards those late writes once the timeout path wins. The client-visible body comes from the timeout middleware itself, which returns the standard `Request Timeout` response. I verified that behavior both in this repo and in the `goravel/example` integration test.

Comment (hwbrzzl) [reply middleware_timeout_test.go:81]: Hence, `return ctx.Response().Status(contractshttp.StatusRequestTimeout).String("timeout")` is no effect, correct? It will be abandoned.

File: context.go [hunk 6]
```diff
@@ -99,15 +104,15 @@ func (c *Context) Context() context.Context {
 }
 
 func (c *Context) Deadline() (deadline time.Time, ok bool) {
-	return c.instance.Deadline()
+	return c.instance.Request.Context().Deadline()
```
Comment (hwbrzzl) [inline context.go:109]: Why is `Request.Context()` required? Why not `c.instance.Deadline()` directly?

File: middleware_timeout_test.go (hunk 2 above)
Comment (hwbrzzl) [reply middleware_timeout_test.go:81]: Correct. In this test handler, `return ctx.Response().Status(contractshttp.StatusRequestTimeout).String("timeout")` is intentionally a late write after `<-ctx.Done()`. `gin-contrib/timeout` has already chosen the timeout path at that point, so that write is discarded. The test uses that shape on purpose to prove late writes cannot leak into the next request.

File: context.go (hunk 6 above)
Comment (hwbrzzl) [reply context.go:109]: Because `c.instance.Deadline()/Done()/Err()` only reflect `Request.Context()` when Gin's `Engine.ContextWithFallback` is enabled. This driver does not enable that flag, so reading through `Request.Context()` is the reliable way to expose the timeout context that the middleware installs with `WithContext(...)`.

Comment (hwbrzzl) [reply context.go:109]: If so, is it reasonable to set `ContextWithFallback=true` instead of changing ` c.instance.Deadline()` to `c.instance.Request.Context().Deadline()`?

Comment (hwbrzzl) [reply context.go:109]: I do not think `ContextWithFallback=true` is a better fit here. It makes `gin.Context.Deadline()/Done()/Err()/Value()` depend on an engine-level switch, while this wrapper should expose the request context deterministically on its own. It also would not help standalone contexts created outside the main engine path unless every such engine or test context also enables that flag. Reading `Request.Context()` directly keeps the Goravel contract tied to the context installed by `WithContext(...)` without another global Gin precondition.

File: context.go [hunk 7]
```diff
@@ -35,17 +37,20 @@ var contextPool = sync.Pool{New: func() any {
...
 	instance *gin.Context
-	request  http.ContextRequest
-	response http.ContextResponse
+	request  contractshttp.ContextRequest
+	response contractshttp.ContextResponse
 }
 
 func NewContext(c *gin.Context) *Context {
 	ctx := contextPool.Get().(*Context)
+	if c.Request == nil {
```
Comment (hwbrzzl) [inline context.go:48]: Add an annotation to explain the reason here.

Comment (hwbrzzl) [reply context.go:48]: Added a short annotation in `b734b63` explaining that some test-created Gin contexts do not carry a request, while Goravel's accessors always read through `Request.Context()`.
